specification/keyvault/resource-manager/Microsoft.KeyVault/stable/2019-09-01/keys.json

```diff
@@ -0,0 +1,588 @@
+          "List keys in the vault": {
+            "$ref": "./examples/listKeys.json"
+          }
+        },
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
+        },
+        "produces": [
+          "application/json"
+        ],
+        "consumes": [
+          "application/json"
+        ]
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}/keys/{keyName}/versions/{keyVersion}": {
+      "get": {
+        "tags": [
+          "Keys"
+        ],
+        "operationId": "Keys_GetVersion",
+        "description": "Gets the specified version of the specified key in the specified key vault.",
+        "parameters": [
+          {
+            "$ref": "#/parameters/SubscriptionIdParameter"
+          },
+          {
+            "name": "resourceGroupName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "description": "The name of the resource group which contains the specified key vault."
+          },
+          {
+            "name": "vaultName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "pattern": "^[a-zA-Z0-9-]{3,24}$",
+            "description": "The name of the vault which contains the key version to be retrieved."
+          },
+          {
+            "name": "keyName",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "pattern": "^[a-zA-Z0-9-]{1,127}$",
+            "description": "The name of the key version to be retrieved."
+          },
+          {
+            "name": "keyVersion",
+            "in": "path",
+            "required": true,
+            "type": "string",
+            "pattern": "^[a-fA-F0-9]{32}$",
+            "description": "The version of the key to be retrieved."
+          },
+          {
+            "$ref": "#/parameters/ApiVersionParameter"
+          }
+        ],
+        "responses": {
+          "200": {
+            "description": "The retrieved key version.",
+            "schema": {
+              "$ref": "#/definitions/Key"
+            }
+          },
+          "default": {
+            "description": "Error response describing why the operation failed.",
+            "schema": {
+              "$ref": "#/definitions/CloudError"
+            }
+          }
+        },
+        "x-ms-examples": {
+          "Get a key version": {
+            "$ref": "./examples/getKeyVersion.json"
+          }
+        },
+        "produces": [
+          "application/json"
+        ],
+        "consumes": [
+          "application/json"
+        ]
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.KeyVault/vaults/{vaultName}/keys/{keyName}/versions": {
+      "get": {
+        "tags": [
+          "Keys"
+        ],
+        "operationId": "Keys_ListVersions",
```

Comment:
Question: what's the use case for this one? ListKeys I can see being used in an ARM function to query for a specific key (must like getting a storage account key), but the versions themselves are, for all intent to the customer, opaque versions. Referring to a specific key (`Keys_GetVersion`) makes sense to refer to a fixed version, but why support listing them in ARM?

Reply:
Good question. The ListKeys/ListKeyVersions APIs are not intended to be used in ARM templates (they are paginated by continuation tokens, which I don't think ARM templates support). Instead, they are intended to be used by UX in the Azure Portal.

The scenario is that a partner team (e.g.: Azure Disk Encryption) will allow customers to provide customer-managed keys by opening a picker control owned by the Key Vault team. This picker will have a dropdown or grid that will list all keys in a given vault, as well as list all key versions for a given key. This picker will call these APIs via ARM instead of directly accessing the data plane because the two code paths use different permission models:
- **Direct data plane APIs:** Permissions to list keys or key versions are configured either via legacy access policies, or via RBAC DataActions.
- **Data plane APIs proxied via control plane:** Permissions to list keys or key versions are configured via RBAC Actions, and can bypass firewall.

The reasoning behind using a separate permission model for these new APIs instead of overloading the existing RBAC DataActions:
- **Fundamentally different APIs:** Although both accomplish the same thing, CreateKey Action is a management operation (typically, a one-time only setup). The CreateKey DataAction is a data operation (used frequently).
- **Restricted elevation of privilege:** We want users with permissions to create vaults via ARM to implicitly also have permission to indirectly create keys via ARM. This happens because owners/contributors have permissions that match `Microsoft.KeyVault/vaults/*` RBAC Actions, which includes the CreateKey Action. However, we don't want these users to automatically be able to create keys directly via the data plane.